bigquery_etl/docs/generate_docs.py

```diff
@@ -0,0 +1,147 @@
+"""Generates documentation for a project."""
+
+from argparse import ArgumentParser
+import os
+from pathlib import Path
+import re
+import shutil
+import yaml
+
+from bigquery_etl.util import standard_args
+
+DEFAULT_PROJECTS = ["mozfun/"]
+DOCS_FILE = "README.md"
+DOCS_DIR = "docs/"
+INDEX_MD = "index.md"
+SQL_REF_RE = r"@sql\((.+)\)"
+
+parser = ArgumentParser(description=__doc__)
+parser.add_argument(
+    "--project_dirs",
+    "--project-dirs",
+    help="Directories of projects documentation is generated for.",
+    nargs="+",
+    default=DEFAULT_PROJECTS,
+)
+parser.add_argument(
+    "--docs_dir",
+    "--docs-dir",
+    default=DOCS_DIR,
+    help="Directory containing static documentation.",
+)
+parser.add_argument(
+    "--output_dir",
+    "--output-dir",
+    required=True,
+    help="Generated documentation is written to this output directory.",
+)
+standard_args.add_log_level(parser)
+
+
+def load_with_examples(file):
+    """Load doc file and replace SQL references with examples."""
+    with open(file) as doc_file:
+        file_content = doc_file.read()
+
+        path, _ = os.path.split(file)
+
+        for sql_ref in re.findall(SQL_REF_RE, file_content):
+            sql_example_file = path / Path(sql_ref)
+            with open(sql_example_file) as example_sql:
+                md_sql = f"```sql\n{example_sql.read()}\n```"
+                file_content = file_content.replace(f"@sql({sql_ref})", md_sql)
+
+    return file_content
+
+
+def generate_docs(out_dir, project_dirs, mkdocs_file):
+    """Generate documentation for project and returns mkdocs config."""
+
+    with open(mkdocs_file, "r") as yaml_stream:
+        mkdocs = yaml.safe_load(yaml_stream)
+
+        dir_structure = {}
+
+        for project_dir in project_dirs:
+            if os.path.isdir(project_dir):
+                for root, dirs, files in os.walk(project_dir):
+                    if DOCS_FILE in files:
+                        Path(os.path.join(out_dir, root)).mkdir(
+                            parents=True, exist_ok=True
+                        )
+
+                        # copy doc file to output and replace example references
+                        src = os.path.join(root, DOCS_FILE)
+                        dest = Path(os.path.join(out_dir, root)) / INDEX_MD
+                        dest.write_text(load_with_examples(src))
+
+                        # parse the doc directory structure
+                        # used in mkdocs.yml
+                        path_parts = root.split(os.sep)
+                        config = dir_structure
+
+                        for part in path_parts:
+                            config = config.setdefault(part, {})
+
+    # convert directory structure to mkdocs compatible format
```

Comment:
Generating the `mkdocs.yml` based on the directory structure is where it gets hairy :(
In some cases, each UDF might have its own documentation in addition to the dataset documentation while in some cases there might just be documentation on a dataset level:

```yaml
nav:
  - project_name:
     - dataset_name1: 
        - Overview: path/to/index.md
        - udf1: path/to/index.md
        - udf2: path/to/index.md
      - dataset_name2: path/to/index.md
```

Reply:
Turns out mkdocs auto-generates navigation when not specified. So this is not needed.